Variable sets

rexredinger opened the pull request:
<!--
Thank you for contributing to hashicorp/go-tfe!

Here's what to expect after the pull request is opened:

The test suite contains many acceptance tests that are run against a test version of Terraform Cloud and additional testing is done against Terraform Enterprise. You can read more about running the tests against your own Terraform Enterprise environment in TESTS.md. Our CI system (Circle) will not test your fork unless you are an authorized employee, so a HashiCorp maintainer will initiate the tests or you and report any missing tests or simple problems. In order to speed up this process, it's not uncommon for your commits to be incorportated into another PR that we can commit test changes to.

Your change, depending on its impact, may be released in the following ways:

  1. For impactful bug fixes, it can be released fairly quickly as a patch release
  2. For noncritical bug fixes and new features, it will be incorporated into the next minor version release
  3. For breaking changes (those changes that alter the public method signatures), more consideration must be made and alternatives may be considered, depending on upgrade difficulty and release schedule.

Please fill out the remaining template to assist code reviewers and testers with incorporating your change. If a section does not apply, feel free to delete it.
-->

## Description

Fixes #284

This adds functionality for managing Variable Sets and their Variables. The plan is 

## Testing plan

1. CRUD a Variable Set.
2. CRUD some Variables against a Variable Set.
3. Assign a Variable Set to some workspaces.

## External links

<!--
_Include any links here that might be helpful for people reviewing your PR. If there are none, feel free to delete this section._

- [API documentation](https://www.terraform.io/cloud-docs/api-docs/variable-sets)
- Related Terraform Provider PR soon to come.

-->

## Output from tests (HashiCorp employees only)

<!--
_Please run the tests locally for any files you changes and include the output here._
-->

NOTE: TFE Tests are expected to fail because of extant feature flagging.

```
🍔_alex: go-tfe $ envchain gotfe go test -run "^TestVariableSet.+" -v ./... -tags=integration
=== RUN   TestVariableSetsList
=== RUN   TestVariableSetsList/without_list_options
    variable_set_test.go:32: paging not supported yet in API
=== RUN   TestVariableSetsList/with_list_options
    variable_set_test.go:38: paging not supported yet in API
=== RUN   TestVariableSetsList/when_Organization_name_is_invalid_ID
--- PASS: TestVariableSetsList (1.53s)
    --- SKIP: TestVariableSetsList/without_list_options (0.19s)
    --- SKIP: TestVariableSetsList/with_list_options (0.00s)
    --- PASS: TestVariableSetsList/when_Organization_name_is_invalid_ID (0.00s)
=== RUN   TestVariableSetsCreate
=== RUN   TestVariableSetsCreate/with_valid_options
=== RUN   TestVariableSetsCreate/when_options_is_missing_name
=== RUN   TestVariableSetsCreate/when_options_is_missing_global_flag
--- PASS: TestVariableSetsCreate (0.98s)
    --- PASS: TestVariableSetsCreate/with_valid_options (0.20s)
    --- PASS: TestVariableSetsCreate/when_options_is_missing_name (0.00s)
    --- PASS: TestVariableSetsCreate/when_options_is_missing_global_flag (0.00s)
=== RUN   TestVariableSetsRead
=== RUN   TestVariableSetsRead/when_the_variable_set_exists
=== RUN   TestVariableSetsRead/when_variable_set_does_not_exist
--- PASS: TestVariableSetsRead (1.27s)
    --- PASS: TestVariableSetsRead/when_the_variable_set_exists (0.10s)
    --- PASS: TestVariableSetsRead/when_variable_set_does_not_exist (0.19s)
=== RUN   TestVariableSetsUpdate
=== RUN   TestVariableSetsUpdate/when_updating_a_subset_of_values
=== RUN   TestVariableSetsUpdate/when_options_has_an_invalid_variable_set_ID
--- PASS: TestVariableSetsUpdate (0.92s)
    --- PASS: TestVariableSetsUpdate/when_updating_a_subset_of_values (0.10s)
    --- PASS: TestVariableSetsUpdate/when_options_has_an_invalid_variable_set_ID (0.00s)
=== RUN   TestVariableSetsDelete
=== RUN   TestVariableSetsDelete/with_valid_ID
=== RUN   TestVariableSetsDelete/when_ID_is_invlaid
--- PASS: TestVariableSetsDelete (0.99s)
    --- PASS: TestVariableSetsDelete/with_valid_ID (0.19s)
    --- PASS: TestVariableSetsDelete/when_ID_is_invlaid (0.00s)
=== RUN   TestVariableSetsAssign
=== RUN   TestVariableSetsAssign/with_valid_workspaces
--- PASS: TestVariableSetsAssign (1.36s)
    --- PASS: TestVariableSetsAssign/with_valid_workspaces (0.33s)
=== RUN   TestVariableSetVariablesList
=== RUN   TestVariableSetVariablesList/without_list_options
    variable_set_variable_test.go:39: paging not supported yet in API
=== RUN   TestVariableSetVariablesList/with_list_options
    variable_set_variable_test.go:45: paging not supported yet in API
=== RUN   TestVariableSetVariablesList/when_variable_set_ID_is_invalid_ID
--- PASS: TestVariableSetVariablesList (1.40s)
    --- SKIP: TestVariableSetVariablesList/without_list_options (0.17s)
    --- SKIP: TestVariableSetVariablesList/with_list_options (0.00s)
    --- PASS: TestVariableSetVariablesList/when_variable_set_ID_is_invalid_ID (0.00s)
=== RUN   TestVariableSetVariablesCreate
=== RUN   TestVariableSetVariablesCreate/with_valid_options
=== RUN   TestVariableSetVariablesCreate/when_options_has_an_empty_string_value
=== RUN   TestVariableSetVariablesCreate/when_options_has_an_empty_string_description
=== RUN   TestVariableSetVariablesCreate/when_options_has_a_too-long_description
=== RUN   TestVariableSetVariablesCreate/when_options_is_missing_value
=== RUN   TestVariableSetVariablesCreate/when_options_is_missing_key
=== RUN   TestVariableSetVariablesCreate/when_options_has_an_empty_key
=== RUN   TestVariableSetVariablesCreate/when_options_is_missing_category
=== RUN   TestVariableSetVariablesCreate/when_workspace_ID_is_invalid
--- PASS: TestVariableSetVariablesCreate (1.51s)
    --- PASS: TestVariableSetVariablesCreate/with_valid_options (0.13s)
    --- PASS: TestVariableSetVariablesCreate/when_options_has_an_empty_string_value (0.14s)
    --- PASS: TestVariableSetVariablesCreate/when_options_has_an_empty_string_description (0.11s)
    --- PASS: TestVariableSetVariablesCreate/when_options_has_a_too-long_description (0.09s)
    --- PASS: TestVariableSetVariablesCreate/when_options_is_missing_value (0.13s)
    --- PASS: TestVariableSetVariablesCreate/when_options_is_missing_key (0.00s)
    --- PASS: TestVariableSetVariablesCreate/when_options_has_an_empty_key (0.00s)
    --- PASS: TestVariableSetVariablesCreate/when_options_is_missing_category (0.00s)
    --- PASS: TestVariableSetVariablesCreate/when_workspace_ID_is_invalid (0.00s)
=== RUN   TestVariableSetVariablesRead
=== RUN   TestVariableSetVariablesRead/when_the_variable_exists
=== RUN   TestVariableSetVariablesRead/when_the_variable_does_not_exist
=== RUN   TestVariableSetVariablesRead/without_a_valid_variable_set_ID
=== RUN   TestVariableSetVariablesRead/without_a_valid_variable_ID
--- PASS: TestVariableSetVariablesRead (1.26s)
    --- PASS: TestVariableSetVariablesRead/when_the_variable_exists (0.13s)
    --- PASS: TestVariableSetVariablesRead/when_the_variable_does_not_exist (0.09s)
    --- PASS: TestVariableSetVariablesRead/without_a_valid_variable_set_ID (0.00s)
    --- PASS: TestVariableSetVariablesRead/without_a_valid_variable_ID (0.00s)
=== RUN   TestVariableSetVariablesUpdate
=== RUN   TestVariableSetVariablesUpdate/with_valid_options
=== RUN   TestVariableSetVariablesUpdate/when_updating_a_subset_of_values
=== RUN   TestVariableSetVariablesUpdate/with_sensitive_set
=== RUN   TestVariableSetVariablesUpdate/without_any_changes
=== RUN   TestVariableSetVariablesUpdate/with_invalid_variable_ID
=== RUN   TestVariableSetVariablesUpdate/with_invalid_variable_ID#01
--- PASS: TestVariableSetVariablesUpdate (1.66s)
    --- PASS: TestVariableSetVariablesUpdate/with_valid_options (0.11s)
    --- PASS: TestVariableSetVariablesUpdate/when_updating_a_subset_of_values (0.10s)
    --- PASS: TestVariableSetVariablesUpdate/with_sensitive_set (0.11s)
    --- PASS: TestVariableSetVariablesUpdate/without_any_changes (0.31s)
    --- PASS: TestVariableSetVariablesUpdate/with_invalid_variable_ID (0.00s)
    --- PASS: TestVariableSetVariablesUpdate/with_invalid_variable_ID#01 (0.00s)
=== RUN   TestVariableSetVariablesDelete
=== RUN   TestVariableSetVariablesDelete/with_valid_options
=== RUN   TestVariableSetVariablesDelete/with_non_existing_variable_ID
=== RUN   TestVariableSetVariablesDelete/with_invalid_workspace_ID
=== RUN   TestVariableSetVariablesDelete/with_invalid_variable_ID
--- PASS: TestVariableSetVariablesDelete (1.19s)
    --- PASS: TestVariableSetVariablesDelete/with_valid_options (0.13s)
    --- PASS: TestVariableSetVariablesDelete/with_non_existing_variable_ID (0.09s)
    --- PASS: TestVariableSetVariablesDelete/with_invalid_workspace_ID (0.00s)
    --- PASS: TestVariableSetVariablesDelete/with_invalid_variable_ID (0.00s)
PASS
ok  	github.com/hashicorp/go-tfe	17.820s
?   	github.com/hashicorp/go-tfe/examples/organizations	[no test files]
?   	github.com/hashicorp/go-tfe/examples/workspaces	[no test files]
?   	github.com/hashicorp/go-tfe/mocks	[no test files]
```

brandonc commented on variable_set.go:
```diff
@@ -0,0 +1,255 @@
+}
+
+// List all Variable Sets in the organization
+func (s *variableSets) List(ctx context.Context, organization string, options VariableSetListOptions) (*VariableSetList, error) {
```
In the [1.0 release](https://github.com/hashicorp/go-tfe/pull/291), all list options will be expressed as a pointer to simplify callers who provide default list options. Maybe we can get ahead of that now?

brandonc commented on variable_set.go:
```diff
@@ -0,0 +1,255 @@
+	Name *string `jsonapi:"attr,name"`
+
+	// A description to provide context for the variable set.
+	Description *string `jsonapi:"attr,description"`
+
+	// If true the variable set is considered in all runs in the organization.
+	Global *bool `jsonapi:"attr,global"`
```
Since these two are not required by the API, I think they should be annotated omitempty so that null isn't passed to the API. I also noticed that you required Global in line 113 so that null wouldn't be passed but the API doesn't have the same requirement and defaults to false. If this were omitempty, I don't think you'd have to make that nil check

rexredinger replied:
Good call!

brandonc commented on variable_set.go:
```diff
@@ -0,0 +1,255 @@
+		return nil, errors.New("invalid variable set ID")
+	}
+
+	u := fmt.Sprintf("varsets/%s?include=workspaces", url.QueryEscape(variableSetID))
```
Usually we allow users to specify include params using ReadOptions and ListOptions, which populates the relation fields in the struct. In the 1.0 release we'll be [making these go typed strings](https://github.com/hashicorp/go-tfe/pull/294)

brandonc commented on variable_set.go:
```diff
@@ -0,0 +1,255 @@
+	// https://www.terraform.io/cloud-docs/api-docs/variable-sets#apply-variable-set-to-workspaces
+	Name *string `jsonapi:"attr,name"`
+
+	// A description to provide context for the variable set.
+	Description *string `jsonapi:"attr,description"`
+
+	// If true the variable set is considered in all runs in the organization.
+	Global *bool `jsonapi:"attr,global"`
```
these may need to be omitempty or else won't rails try to clear these fields if they aren't set?

brandonc commented on variable_set.go:
```diff
@@ -0,0 +1,255 @@
+		return nil, errors.New("invalid value for variable set ID")
+	}
+
+	u := fmt.Sprintf("varsets/%s?include=workspaces", url.QueryEscape(variableSetID))
```
☝️ forces include=workspaces

brandonc commented on variable_set.go:
```diff
@@ -0,0 +1,255 @@
+		return nil, errors.New("No workspaces list provided")
+	}
+
+	options.Global = Bool(false)
```
Maybe this should not be an attribute on the assign options since you set it statically? Is it necessary to always pass false to this endpoint?

rexredinger replied:
For this case, Yes. The implication that you want to assign specific workspaces means you no longer want the variable set to be global.

If not set to false, and the variable set is global, workspace assignment will fail.

brandonc reviewed (changes requested):
A lot of my comments were made on VariableSets but apply to VariableSetVariables as well and I didn't point them out explicitly. I noticed many requests implicitly included workspace data. Is that necessary in a way that I overlooked?

rexredinger commented:
> I noticed many requests implicitly included workspace data. Is that necessary in a way that I overlooked?

Not required in all cases, but will be needed or useful in some workflows. I wasn't aware of the ReadOptions pattern, though. That seems better. Thanks for the review!

brandonc commented:
That is a lot of failures! Something may be up with our nightly TFE environment? I'll look into it because I'm investigating the test flakes anyway

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/go-tfe?pullRequest=305)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/go-tfe?pullRequest=305)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/go-tfe?pullRequest=305).</sub>

rexredinger commented:
I merged main and the tests exploded. 😭

annawinkler commented:
Is there an associated Asana issue or other documentation for this work?

nfagerlund commented:
@rexredinger (I've edited your description to add a "fixes" line, since it looks like there's an existing issue.)

nfagerlund reviewed (changes requested):
Hey Alex!

- First off: this probably needs a rebase onto `main` before going much further -- 1.0 happened, and some of the changes we're asking for will be easier when based on current.

- In `VariableSets.Assign()`, it's not clear whether you implemented the behavior you wanted!

    We dug up two ways to modify the workspace relationship: there's PATCH on the whole varset resource while setting the relationships object, which _replaces_ the entire set of assigned workspaces. And there's also POST on `/varsets/:id/relationships/workspaces`, which _adds_ a new varset/workspace relationship without erasing the existing ones.

    The godoc comment on Assign() seems to imply that it's just assigning the varset to an additional workspace, but really it's replacing the entire set. 

    If you want it to behave like the comment says, you'll need to switch the endpoint you're hitting, and you might also want to implement the `DELETE /varsets/:id/relationships/workspaces` endpoint in a separate `Unassign()` function if you're doing that. 
    
    If you definitely want the existing behavior... maybe remove Assign entirely and document the use of Update to do the same thing? We're not sure! Alternately, you could update the comment on Assign to explain what's about to happen.

    (One thing to be aware of if you go the additive `POST .../relationships` route is that there's currently a nasty unhandled 500 error in the API if you try to additively modify the relationships while a workspace is set to `global: true`. So, potentially two API requests necessary to avoid that.)

- The VariableSetUpdateOptions struct doesn't have `omitempty` on Name, so it's impossible to e.g. change a varset to global without also passing a name! Please fix that, and maybe check over the other Options structs to make sure everything that logically ought to be `omitempty` is.

- Luces spotted that the error handling in this PR is now out-of-step with what the rest of the code does (part of the 1.0 stuff) -- instead of calling `errors.New()`, options validation errors should be referencing stuff from errors.go now.

rexredinger commented:
Thanks for the feedback!

@annawinkler We have an asana project, but not a particular task or ticket that will shed additional light on the work here. Related though is the [docs draft branch](https://github.com/hashicorp/terraform-provider-tfe/tree/variable-reuse/docs-draft) for the TFE provider work that would consume this.

@nfagerlund  I rebased, updated the errors, and added omitempty to the update options. As for Assign clarity, I wanted to take advantage of the idempotence of the Update path for workspace assignment, but narrow down the interface to allow for assignment without the risk of other mutation. Happy to move all responsibility to Update if you think that's more appropriate. I'm also happy to just update doc verbiage to be clearer about Assign being both destructive and idempotent.

nfagerlund reviewed (commented):
This is still WIP, but I have a couple comments that are independent of whatever we decide about Assign(), so I'm going to go ahead and post those while the conversation is in progress!

rexredinger commented on variable_integration_test.go:
```diff
@@ -128,7 +128,7 @@ func TestVariablesCreate(t *testing.T) {
 			Key:         String(randomString(t)),
 			Value:       String(randomString(t)),
-			Description: String("tortor aliquam nulla facilisi cras fermentum odio eu feugiat pretium nibh ipsum consequat nisl vel pretium lectus quam id leo in vitae turpis massa sed elementum tempus egestas sed sed risus pretium quam vulputate dignissim suspendisse in est ante in nibh mauris cursus mattis molestie a iaculis at erat pellentesque adipiscing commodo elit at imperdiet dui accumsan sit amet ...
+			Description: String("tortor aliquam nulla go lint is fussy about spelling cras fermentum odio eu feugiat pretium nibh ipsum consequat nisl vel pretium lectus quam id leo in vitae turpis massa sed elementum tempus egestas sed sed risus pretium quam vulputate dignissim suspendisse in est ante in nibh mauris cursus mattis molestie a iaculis at erat pellentesque adipiscing commodo elit at imperdie ...
```
go lint error:
"Perhaps you meant 'faciltities'"

rexredinger replied:
`"iaculis" checks out though`

rexredinger commented on variable_set.go:
```diff
@@ -0,0 +1,276 @@
+	// The name of the variable set.
+	// Affects variable precedence when there are conflicts between Variable Sets
+	// https://www.terraform.io/cloud-docs/api-docs/variable-sets#apply-variable-set-to-workspaces
+	Name *string `jsonapi:"attr,name"`
```
AH! I swear I updated this. Maybe something got mangled in the rebase. Thanks for catching it.

rexredinger commented on variable_set.go:
```diff
@@ -0,0 +1,276 @@
+	// If true the variable set is considered in all runs in the organization.
+	Global *bool `jsonapi:"attr,global,omitempty"`
+
+	Include *[]VariableSetIncludeOps `url:"include:omitempty"`
```
I can remove this. I was erring more on the side of "completeness", but this is never used or needed.

uturunku1 commented on variable_set.go:
```diff
@@ -0,0 +1,274 @@
+	if !validStringID(&organization) {
+		return nil, ErrInvalidOrg
+	}
+	if options != nil {
```
Can this check be moved to the valid() helper function? Like this: https://github.com/hashicorp/go-tfe/blob/20889e373e12cd0f20c35d7df3a37cfb711c7d70/admin_organization.go#L231

nfagerlund reviewed (changes requested):
@rexredinger Ok, thank you for your patience with all the back-and-forth! 

I consulted with the team about how to handle these kinds of relationship updates consistently and idiomatically, and we agreed that the current gold standard for this kind of behavior is over in Workspaces, on the RemoteStateConsumers relationship: 

https://github.com/hashicorp/go-tfe/blob/716b4e46f31bbd22acbb8d1f9eb099c084d279c9/workspace.go#L928-L979

In summary, it uses three methods: `UpdateRemoteStateConsumers` for replacing all, and `AddRemoteStateConsumers` / `RemoveRemoteStateConsumers` for in-place modifications. 

For this PR, we want to make sure that varsets are at least on track for being consistent with that approach. That means: 

- At minimum, we'd like `Apply` renamed to `UpdateWorkspaces` -- consistency with the rest of go-tfe is more important in this context than consistency with the UI for that feature. (Yes, go-tfe is already inconsistent, cf. other things like PolicySets that are almost exactly like this thing... still though!!)
- If you can tolerate the extra work of adding the in-place add/remove functions, we would truly appreciate it! But we also know you're dealing with a lot of dependencies at the moment on this feature, so it's OK if you need to mark them as TODOs and create a follow-up ticket instead. 
- I'd like the (to-be-renamed) public `VariableSetUpdateWorkspacesOptions` struct to NOT include the Global field. The user can never set it and it's just an implementation detail to work around an odd interaction in the API, but it'll show up as clutter in the go-docs for that struct. I think the move is probably to have a private struct type (that includes Global) that you _actually_ pass to `client.newRequest()`, and transfer the Workspaces slice from the public struct argument into that.

rexredinger commented:
@uturunku1 Thanks!

@nfagerlund
> Ok, thank you for your patience with all the back-and-forth!
Thanks for being patient with _me_ 😅 

* I can and will definitely do the minimum of renaming to `UpdateWorkspaces`
* I would love to jump on updating the API to support these options, but I just do not have the bandwidth at this juncture. But, rest assured, follow-up tickets will be made and followed-up on.
* Will do!

nfagerlund commented on variable_set.go:
```diff
@@ -104,7 +104,7 @@ type VariableSetCreateOptions struct {
 	// It is not a user-defined value and does not need to be set.
 	// https://jsonapi.org/format/#crud-creating
-	Type string `jsonapi:"primary,vars"`
+	Type string `jsonapi:"primary,varsets"`
```
!! good catch. How did I not see that earlier?!

nfagerlund commented on variable_set.go:
```diff
@@ -0,0 +1,289 @@
+		return nil, err
+	}
+
+	// Use private strcut to ensure global is set to false when applying to workspaces
```
```suggestion
	// Use private struct to ensure global is set to false when applying to workspaces
```

nfagerlund commented on variable_set_test.go:
```diff
@@ -0,0 +1,224 @@
+		assert.Equal(t, ErrResourceNotFound, err)
+	})
+
+	t.Run("when ID is invlaid", func(t *testing.T) {
```
```suggestion
	t.Run("when ID is invalid", func(t *testing.T) {
```

nfagerlund commented on variable_set_test.go:
```diff
@@ -0,0 +1,224 @@
+	defer orgTestCleanup()
+
+	vsTest, _ := createVariableSet(t, client, orgTest, VariableSetCreateOptions{
+		Name:        String("OrinigalName"),
```
```suggestion
		Name:        String("OriginalName"),
```

rexredinger replied:
I literally struggle to see the problem even with side by side diffs 🤣

nfagerlund reviewed (approved):
Okay, I think this PR is good to go! Thank you for the revisions! 

I spotted a few trivial typos in comments and in test data strings, but I'm not going to block the PR on them; you're free to merge. If you do decide to tweak them, ping me in Slack and I'll quickfast re-approve.

sebasslash reviewed (approved):
Double approve 🔥
